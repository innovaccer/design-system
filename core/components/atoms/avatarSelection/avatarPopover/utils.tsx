import * as React from 'react';

export const handleKeyDown = (
  event: React.KeyboardEvent,
  focusedOption: Element | undefined,
  setFocusedOption?: React.Dispatch<React.SetStateAction<HTMLElement | undefined>>,
  setHighlightFirstItem?: React.Dispatch<React.SetStateAction<boolean>>,
  setHighlightLastItem?: React.Dispatch<React.SetStateAction<boolean>>,
  listRef?: any,
  withSearch?: boolean,
  setOpenPopover?: React.Dispatch<React.SetStateAction<boolean>>,
  triggerRef?: any
) => {
  switch (event.key) {
    case 'ArrowUp':
      event.preventDefault();
      navigateOptions('up', focusedOption, setFocusedOption, listRef, withSearch);
      break;
    case 'ArrowDown':
      event.preventDefault();
      navigateOptions('down', focusedOption, setFocusedOption, listRef, withSearch);
      break;
    case 'Enter':
      handleEnterKey(focusedOption);
      setHighlightLastItem?.(false);
      setHighlightFirstItem?.(false);
      break;
    case 'Escape':
      setOpenPopover?.(false);
      triggerRef.current.focus();
      setFocusedOption?.(undefined);
      break;
    default:
      break;
  }
};

const handleEnterKey = (focusedOption: Element | undefined) => {
  (focusedOption as HTMLElement)?.click();
};

const navigateOptions = (
  direction: string,
  focusedOption: Element | undefined,
  setFocusedOption?: React.Dispatch<React.SetStateAction<HTMLElement | undefined>>,
  listRef?: any,
  withSearch?: boolean
) => {
  const listItems = listRef?.current.querySelectorAll('[data-test="DesignSystem-Listbox-ItemWrapper"]');
  let index = Array.from(listItems).findIndex((item) => {
    return item == focusedOption;
  });

  if (index === -1) {
    index = direction === 'up' ? listItems.length - 1 : 0;
  } else if (
    (withSearch && index === 0 && direction === 'up') ||
    (withSearch && index === listItems.length - 1 && direction === 'down')
  ) {
    const searchInput = listRef.current.querySelector('[data-test="DesignSystem-AvatarSelection--Input"]');
    searchInput.focus();
    setFocusedOption && setFocusedOption(searchInput);
  } else {
    index = direction === 'up' ? (index - 1 + listItems.length) % listItems.length : (index + 1) % listItems.length;

    const targetOption = listItems[index];

    (targetOption as HTMLElement).focus();
    setFocusedOption && setFocusedOption(targetOption);
    targetOption.scrollIntoView({ block: 'center' });
  }
};

export const handleInputKeyDown = (
  event: React.KeyboardEvent,
  listRef: any,
  setFocusedOption?: React.Dispatch<React.SetStateAction<HTMLElement | undefined>>,
  setOpenPopover?: React.Dispatch<React.SetStateAction<boolean>>,
  triggerRef?: any
) => {
  const listItems = listRef.current?.querySelectorAll('[data-test="DesignSystem-Listbox-ItemWrapper"]');
  let targetOption;

  switch (event.key) {
    case 'ArrowUp':
      event.preventDefault();
      targetOption = listItems[listItems.length - 1];
      break;
    case 'ArrowDown':
      event.preventDefault();
      targetOption = listItems[0];
      break;
    case 'Escape':
      setOpenPopover?.(false);
      triggerRef.current.focus();
      setFocusedOption?.(undefined);
      break;
    default:
      break;
  }

  (targetOption as HTMLElement)?.focus();
  targetOption?.scrollIntoView?.({ block: 'center' });
  setFocusedOption && setFocusedOption(targetOption);
};
