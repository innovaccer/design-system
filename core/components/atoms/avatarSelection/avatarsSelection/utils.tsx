import * as React from 'react';

export const handleKeyDown = (
  event: React.KeyboardEvent,
  setOpenPopover?: React.Dispatch<React.SetStateAction<boolean>>,
  setHighlightFirstItem?: React.Dispatch<React.SetStateAction<boolean>>,
  setHighlightLastItem?: React.Dispatch<React.SetStateAction<boolean>>
) => {
  switch (event.key) {
    case 'Enter':
      setOpenPopover?.(true);
      setHighlightFirstItem?.(true);
      break;
    case 'ArrowDown':
      setOpenPopover?.(true);
      setHighlightFirstItem?.(true);
      break;
    case 'ArrowUp':
      setOpenPopover?.(true);
      setHighlightLastItem?.(true);
      break;
    default:
      break;
  }
};

export const focusListItem = (
  position: string,
  setFocusedOption?: React.Dispatch<React.SetStateAction<HTMLElement | undefined>>,
  listRef?: any,
  withSearch?: boolean
) => {
  if (withSearch && position === 'down') {
    const searchInput = listRef.current?.querySelectorAll('[data-test="DesignSystem-AvatarSelection--Input"]');
    searchInput && searchInput[0]?.focus();
    setFocusedOption?.(searchInput);
    return;
  }

  const listItems = listRef.current?.querySelectorAll('[data-test="DesignSystem-Listbox-ItemWrapper"]');
  let targetOption;

  if (position === 'down') {
    targetOption = listItems?.[0];
  } else {
    targetOption = listItems?.[listItems.length - 1];
  }
  (targetOption as HTMLElement)?.focus();
  targetOption?.scrollIntoView?.({ block: 'center' });
  setFocusedOption && setFocusedOption(targetOption);
};
