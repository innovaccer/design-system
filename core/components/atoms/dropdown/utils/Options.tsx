export let dropdownOptions: any[] = [];

for (let i = 1; i <= 50; i++) {
  dropdownOptions.push({
    label: `Option${i}`,
    value: `Option${i}`,
  });
}

export const selectedStoryOptions: any[] = [];

for (let i = 1; i <= 10; i++) {
  selectedStoryOptions.push({
    label: `Option${i}`,
    value: `Option${i}`,
    selected: i === 2,
  });
}

export const multiSelectedStoryOptions: any[] = [];

for (let i = 1; i <= 10; i++) {
  multiSelectedStoryOptions.push({
    label: `Option${i}`,
    value: `Option${i}`,
    selected: i === 2 || i === 7,
  });
}

export const storyOptions = dropdownOptions.slice(0, 10);

export const storySections = { 0: 'subheading', 5: 'subheading' };

export const subInfoItems = [
  {
    label: 'Option1',
    value: 'Option1',
    subInfo: 'subInfo'
  },
  {
    label: 'Option2',
    value: 'Option2',
    subInfo: 'subInfo'
  },
  {
    label: 'Option3',
    value: 'Option3',
    subInfo: 'subInfo',
  },
];

export const iconItems = [
  {
    label: 'Option1',
    value: 'Option1',
    icon: 'events'
  },
  {
    label: 'Option2',
    value: 'Option2',
    icon: 'events',
  },
  {
    label: 'Option3',
    value: 'Option3',
    icon: 'events',
  },
];
