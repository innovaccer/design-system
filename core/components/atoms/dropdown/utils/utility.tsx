const getSearchedOptions = (options: any, searchTerm: string) => {
  const result = options.filter((option: any) => option.label.toLowerCase().includes(searchTerm.toLowerCase()));
  return result;
};

export const getOptions = (offset: number, limit: number, searchTerm: string, options: any) => {
  const searchedOptions = searchTerm ? getSearchedOptions(options, searchTerm) : options;
  return new Promise(resolve => {
    resolve({
      offset,
      slicedOptions: searchedOptions.slice(offset, offset + limit),
      options: searchedOptions,
    });
  });
};

export const getValuesFromSelectedObj = (selectedArray: any[] = []) => {
  const selectedValues: any[] = [];

  selectedArray.forEach((selectedObj: any) => {
    const { value } = selectedObj;
    selectedValues.push(value);
  });

  return selectedValues;
};

export const getIndexesFromSelectedObj = (selectedArray: any[] = []) => {
  return (Array.from(Array(selectedArray.length).keys()));
};

export const getValuesFromInd = (selectedArray: number[], options: any[]) => {
  const result: any[] = [];
  selectedArray.forEach(ind => {
    result.push(options[ind].value);
  });

  return result;
};

export const getLabelsFromInd = (selectedArray: number[], options: any[]) => {
  const result: string[] = [];
  selectedArray.forEach(ind => {
    result.push(options[ind].label);
  });

  return result;
};
