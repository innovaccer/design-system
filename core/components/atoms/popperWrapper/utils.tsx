// Helper function to convert CSS class names to style objects
export const createStyleFromClass = (className: string) => {
  if (!className) return {};

  // Create a temporary element to get computed styles
  const tempElement = document.createElement('div');
  tempElement.className = className;
  document.body.appendChild(tempElement);

  const computedStyle = window.getComputedStyle(tempElement);
  const style: React.CSSProperties = {};

  // Extract relevant animation properties
  const animation = computedStyle.animation;
  const animationDuration = computedStyle.animationDuration;
  const animationTimingFunction = computedStyle.animationTimingFunction;
  const animationFillMode = computedStyle.animationFillMode;
  const opacity = computedStyle.opacity;
  const transform = computedStyle.transform;

  if (animation && animation !== 'none') {
    style.animation = animation;
  }
  if (animationDuration && animationDuration !== '0s') {
    style.animationDuration = animationDuration;
  }
  if (animationTimingFunction && animationTimingFunction !== 'ease') {
    style.animationTimingFunction = animationTimingFunction;
  }
  if (animationFillMode && animationFillMode !== 'none') {
    style.animationFillMode = animationFillMode;
  }
  if (opacity && opacity !== '1') {
    style.opacity = opacity;
  }
  if (transform && transform !== 'none') {
    style.transform = transform;
  }

  document.body.removeChild(tempElement);
  return style;
};
