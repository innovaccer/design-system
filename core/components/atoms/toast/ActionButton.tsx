import * as React from 'react';
import classNames from 'classnames';
import Text from '@/components/atoms/text';
import { MessageAppearance } from '@/common.type';
import buttonStyles from '@css/components/button.module.css';
import toastStyles from '@css/components/toast.module.css';

export interface Props {
  appearance: MessageAppearance;
  label: string;
  onClick?: (e: React.MouseEvent) => void;
}

const ActionButton = (props: Props) => {
  const { appearance, label, onClick } = props;

  const buttonClass = classNames({
    [buttonStyles['Button']]: true,
    [buttonStyles['Button--tiny']]: true,
    [toastStyles['Toast-actionButton']]: true,
    [toastStyles[`Toast-actionButton--${appearance}`]]: appearance,
  });

  const onClickHandler = (e: React.MouseEvent) => {
    e.preventDefault();
    if (onClick) onClick(e);
  };

  return (
    <button className={buttonClass} onClick={onClickHandler}>
      <Text appearance="white">{label}</Text>
    </button>
  );
};

ActionButton.displayName = 'ActionButton';
ActionButton.defaultProps = {
  appearance: 'default',
};

export default ActionButton;
