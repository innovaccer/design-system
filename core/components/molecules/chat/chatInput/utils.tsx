import { MessageType } from './ChatInput';

export const positionMentionPopup = (
  textareaRef: React.RefObject<HTMLDivElement>,
  setMentionPosition: React.Dispatch<React.SetStateAction<{ top: number; left: number }>>
) => {
  if (!textareaRef.current) return;

  const selection = window.getSelection();
  if (!selection || !selection.rangeCount) return;

  const range = selection.getRangeAt(0);
  const rect = range.getBoundingClientRect();
  const editorRect = textareaRef.current.getBoundingClientRect();

  // Calculate position based on the cursor position
  let top = rect.bottom + window.scrollY;
  let left = rect.left + window.scrollX;

  // Adjust position if it would go off-screen
  if (top + 150 > window.innerHeight) {
    top = rect.top + window.scrollY - 150;
  }
  if (left + 200 > window.innerWidth) {
    left = window.innerWidth - 200;
  }

  // Adjust position relative to the editor
  top -= editorRect.top + window.scrollY - 12;
  left -= editorRect.left + window.scrollX - 8;

  setMentionPosition({ top, left });
};

export const extractMessageData = (textareaRef: React.RefObject<HTMLDivElement>): MessageType[] => {
  const elements = textareaRef.current?.childNodes;

  return Array.from(elements || []).map((item: any) => {
    if (item?.dataset?.type === 'mention') {
      const mentionValue = JSON.parse(item.dataset.content);
      return { type: 'mention', content: mentionValue?.data, nodeName: item.nodeName };
    }

    // handles new line by pressing enter
    else if (item.innerText === '\n' && item.nodeName === 'DIV') {
      return { type: 'newline', content: '\n', nodeName: item.nodeName };
    }

    // handles line break by pressing shift + enter
    else if (item.nodeName === 'BR') {
      return { type: 'linebreak', content: '<br>', nodeName: item.nodeName };
    }

    return { type: 'text', content: item.innerText || item.textContent, nodeName: item.nodeName };
  });
};
