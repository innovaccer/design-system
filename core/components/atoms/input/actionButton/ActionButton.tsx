import * as React from 'react';
import classNames from 'classnames';
import { BaseProps, extractBaseProps } from '@/utils/types';
import { useAccessibilityProps } from '@/accessibility/utils';

export type ActionButtonType = 'filled' | 'outlined' | 'round' | 'two-tone' | 'sharp';

export interface ActionButtonProps extends BaseProps {
  /**
   * Material icon name
   */
  name?: string;
  /**
   * Size of `ActionButton`
   */
  size: number;
  /**
   * Type of material `ActionButton`
   */
  type?: ActionButtonType;
  /**
   * Handler to be called when icon is clicked
   */
  onClick?: (e: React.MouseEvent<HTMLElement>) => void;
  /**
   * Handler to be called when key is pressed on icon
   */
  onKeyDown?: (e: React.KeyboardEvent<HTMLElement>) => void;
  /**
   * DOM node to be passed as child to the component
   */
  children?: React.ReactNode;
  /**
   * The tabindex global attribute indicates that its element can be focused, and
   * where it participates in sequential keyboard navigation.
   */
  tabIndex?: React.DetailedHTMLProps<React.HTMLAttributes<HTMLElement>, HTMLElement>['tabIndex'];
}

export const ActionButton = (props: ActionButtonProps) => {
  const { className, name, size, children, type } = props;
  const accessibilityProps = useAccessibilityProps(props);

  const baseProps = extractBaseProps(props);

  const iconClass = classNames({
    ['material-icons']: true,
    [`material-icons-${type}`]: type && type !== 'filled',
    ['ActionButton']: true,
    [`${className}`]: className,
    ['ml-4']: true,
  });

  const styles = {
    fontSize: `${size}px`,
    width: `${size}px`,
  };

  // change `children` to {name} after migration
  if (children && React.isValidElement(children)) {
    return (
      <span {...baseProps} className={iconClass} data-test="DesignSystem-Input-ActionButton">
        {children}
      </span>
    );
  }
  return (
    <i
      {...baseProps}
      className={iconClass}
      style={styles}
      {...accessibilityProps}
      data-test="DesignSystem-Input-ActionButton"
    >
      {`${name}_${type}`}
    </i>
  );
};

ActionButton.displayName = 'ActionButton';
ActionButton.defaultProps = {
  size: 16,
  type: 'round',
};

export default ActionButton;
