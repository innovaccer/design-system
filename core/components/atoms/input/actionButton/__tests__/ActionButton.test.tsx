import { fireEvent, render } from '@testing-library/react';
import ActionButton, { ActionButtonProps as Props, ActionButtonType } from '../ActionButton';
import { testHelper, filterUndefined, valueHelper, testMessageHelper } from '@/utils/testHelper';

const size = 16;

const type: ActionButtonType[] = ['outlined', 'rounded'];
const FunctionValue = jest.fn();
const StringValue = 'events';

describe('ActionButton component', () => {
  const mapper: Record<string, any> = {
    name: valueHelper(StringValue, { required: true }),
    size: valueHelper(size, { required: true }),
    onClick: valueHelper(FunctionValue, { required: true }),
  };

  const testFunc = (props: Record<string, any>): void => {
    const attr = filterUndefined(props) as Props;

    it(testMessageHelper(attr), () => {
      const { baseElement } = render(<ActionButton {...attr} />);
      expect(baseElement).toMatchSnapshot();
    });
  };
  testHelper(mapper, testFunc);
});

describe('ActionButton component', () => {
  const mapper: Record<string, any> = {
    name: valueHelper(StringValue, { required: true }),
    size: valueHelper(size, { required: true }),
    type: valueHelper(type, { required: true, iterate: true }),
    onClick: valueHelper(FunctionValue, { required: true }),
  };

  const testFunc = (props: Record<string, any>): void => {
    const attr = filterUndefined(props) as Props;

    it(testMessageHelper(attr), () => {
      const { baseElement } = render(<ActionButton {...attr} />);
      expect(baseElement).toMatchSnapshot();
    });
  };
  testHelper(mapper, testFunc);
});

describe('ActionButton component states', () => {
  it('renders action button with hover state', () => {
    const { getByTestId } = render(<ActionButton name={StringValue} />);
    const iconElement = getByTestId('DesignSystem-Input-ActionButton');
    fireEvent.mouseEnter(iconElement);
    expect(iconElement).toHaveClass('ActionButton');
    expect(iconElement).toHaveStyle(`background-color: var(--secondary);`);
  });

  it('renders action button with active state', () => {
    const { getByTestId } = render(<ActionButton name={StringValue} />);
    const iconElement = getByTestId('DesignSystem-Input-ActionButton');
    fireEvent.pointerDown(iconElement);
    expect(iconElement).toHaveClass('ActionButton');
    expect(iconElement).toHaveStyle(`background-color: var(--secondary-dark);`);
  });
});

describe('ActionButton component', () => {
  it('renders action button with children', () => {
    const { getByTestId } = render(
      <ActionButton name={StringValue}>
        <img
          width="50"
          height="50"
          alt="Innovaccer logo"
          data-test="DesignSystem-Image"
          src="https://design.innovaccer.com/images/withoutType.png"
        />
      </ActionButton>
    );
    expect(getByTestId('DesignSystem-Image')).toBeInTheDocument();
    expect(getByTestId('DesignSystem-Input-ActionButton')).toBeInTheDocument();
    expect(getByTestId('DesignSystem-Input-ActionButton')).toHaveClass('ActionButton');
  });
});
