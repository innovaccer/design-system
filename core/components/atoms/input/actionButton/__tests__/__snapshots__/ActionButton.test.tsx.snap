// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ActionButton component 
  name: "events", size: 16, onClick: "[Function]"
 1`] = `
<body>
  <div>
    <i
      class="material-icons material-icons-round ActionButton ml-4"
      data-test="DesignSystem-Input-ActionButton"
      role="button"
      style="font-size: 16px; width: 16px;"
      tabindex="0"
    >
      events_round
    </i>
  </div>
</body>
`;

exports[`ActionButton component 
  name: "events", size: 16, type: "filled", onClick: "[Function]"
 1`] = `
<body>
  <div>
    <i
      class="material-icons ActionButton ml-4"
      data-test="DesignSystem-Input-ActionButton"
      role="button"
      style="font-size: 16px; width: 16px;"
      tabindex="0"
    >
      events_filled
    </i>
  </div>
</body>
`;

exports[`ActionButton component 
  name: "events", size: 16, type: "outlined", onClick: "[Function]"
 1`] = `
<body>
  <div>
    <i
      class="material-icons material-icons-outlined ActionButton ml-4"
      data-test="DesignSystem-Input-ActionButton"
      role="button"
      style="font-size: 16px; width: 16px;"
      tabindex="0"
    >
      events_outlined
    </i>
  </div>
</body>
`;

exports[`ActionButton component 
  name: "events", size: 16, type: "round", onClick: "[Function]"
 1`] = `
<body>
  <div>
    <i
      class="material-icons material-icons-round ActionButton ml-4"
      data-test="DesignSystem-Input-ActionButton"
      role="button"
      style="font-size: 16px; width: 16px;"
      tabindex="0"
    >
      events_round
    </i>
  </div>
</body>
`;

exports[`ActionButton component 
  name: "events", size: 16, type: "sharp", onClick: "[Function]"
 1`] = `
<body>
  <div>
    <i
      class="material-icons material-icons-sharp ActionButton ml-4"
      data-test="DesignSystem-Input-ActionButton"
      role="button"
      style="font-size: 16px; width: 16px;"
      tabindex="0"
    >
      events_sharp
    </i>
  </div>
</body>
`;

exports[`ActionButton component 
  name: "events", size: 16, type: "two-tone", onClick: "[Function]"
 1`] = `
<body>
  <div>
    <i
      class="material-icons material-icons-two-tone ActionButton ml-4"
      data-test="DesignSystem-Input-ActionButton"
      role="button"
      style="font-size: 16px; width: 16px;"
      tabindex="0"
    >
      events_two-tone
    </i>
  </div>
</body>
`;
